jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/parser/Parser.java

```diff
@@ -44,33 +43,34 @@
     private static final Logger LOG = LoggerFactory.getLogger(Parser.class);
 
     private final ByteBufferPool byteBufferPool;
-    private final Listener listener;
     private final HeaderParser headerParser;
     private final HpackDecoder hpackDecoder;
     private final BodyParser[] bodyParsers;
+    private Listener listener;
     private UnknownBodyParser unknownBodyParser;
     private int maxFrameLength = Frame.DEFAULT_MAX_LENGTH;
     private int maxSettingsKeys = SettingsFrame.DEFAULT_MAX_KEYS;
     private boolean continuation;
     private State state = State.HEADER;
 
-    public Parser(ByteBufferPool byteBufferPool, Listener listener, int maxDynamicTableSize, int maxHeaderSize)
+    public Parser(ByteBufferPool byteBufferPool, int maxDynamicTableSize, int maxHeaderSize)
     {
-        this(byteBufferPool, listener, maxDynamicTableSize, maxHeaderSize, RateControl.NO_RATE_CONTROL);
+        this(byteBufferPool, maxDynamicTableSize, maxHeaderSize, RateControl.NO_RATE_CONTROL);
     }
 
-    public Parser(ByteBufferPool byteBufferPool, Listener listener, int maxDynamicTableSize, int maxHeaderSize, RateControl rateControl)
+    public Parser(ByteBufferPool byteBufferPool, int maxDynamicTableSize, int maxHeaderSize, RateControl rateControl)
     {
         this.byteBufferPool = byteBufferPool;
-        this.listener = listener;
         this.headerParser = new HeaderParser(rateControl == null ? RateControl.NO_RATE_CONTROL : rateControl);
         this.hpackDecoder = new HpackDecoder(maxDynamicTableSize, maxHeaderSize);
         this.bodyParsers = new BodyParser[FrameType.values().length];
     }
 
-    public void init(UnaryOperator<Listener> wrapper)
+    public void init(Listener listener)
```

Comment:
Can you explain why the listener can no longer be injected in the constructor?
It might be better to just pass the listener in with every call to parse?
I'm not really understanding the previous wrapper init either.

Reply:
It is a chicken/egg problem.
We need the parser to build the session, we need the session to build the connection, and the connection is the parser listener that we would need to build the parser.
Therefore we need a once-only setter that it is `init()`.

Before, a parser listener (the session) was passed to the constructor of the parser, but then the connection had to wrap it via `init(Function)` -- it was indeed weird.

Now it's clearer because the contract is that you build the parser, and then you have to call `init()` passing the listener (the connection).

Reply:
OK. I don't love it, but better than the wrapper.